contributors/design-proposals/self-hosted-kubernetes.md

```diff
@@ -0,0 +1,99 @@
+
+## Practical Implementation Overview
+
+This document outlines the current implementation of "self-hosted Kubernetes" installation and upgrade of Kubernetes clusters based on the work that the teams at CoreOS and Google have been doing. The work is motivated by the upstream "Self-hosted Proposal".
+
+The entire system is working today and is used by Bootkube, a Kubernetes Incubator project, and all Tectonic clusters created since July 2016. This document outlines the implementation, not the experience. The experience goal is that users not know all of these details and instead get a working Kubernetes cluster out the other end that can be upgraded using the Kubernetes APIs.
+
+The target audience of this document are others, like [kubeadm](https://github.com/kubernetes/kubernetes/pull/38407), thinking about and building the way forward for install and upgrade of Kubernetes. If you want a higher level demonstration of "Self-Hosted" and the value see this [video and blog](https://coreos.com/blog/self-hosted-kubernetes.html).
+
+
+### Bootkube
+
+Today, the first component of the installation of a self-hosted cluster is [`bootkube`](https://github.com/kubernetes-incubator/bootkube). A kubelet connects to the temporary Kubernetes API server provided by bootkube and is told to deploy the required Kubernetes components, as pods. This diagram shows all of the moving parts:
+
+![](self-hosted-moving-parts.png)
+
+At the end of this process the bootkube can be shut down and the system kubelet will coordinate, through a POSIX lock, to let the self-hosted kubelet take over lifecycle and management of the control plane components. The final cluster state looks like this:
+
+![](self-hosted-final-cluster.png)
+
+There are a few things to note. First, generally, the control components like the API server, etc will be pinned to a set of dedicated control nodes. For security policy, service discovery, and scaling reasons it is easiest to assume that control nodes will always exist on N nodes.
+
+Another challenge is load balancing the API server. Bedrock for the API server will be DNS, TLS, and a load balancer that live off cluster and that load balancer will want to only healthcheck a handful of servers for the API server port liveness probe.
+
+### Bootkube Challenges
+
+This process has a number of moving parts. Most notably the hand off of control from the "host system" to the Kubernetes self-hosted system. And things can go wrong:
+
+1) The self-hosted Kubelet is in a precarious position as there is no one around to restart the process if it crashes. The high level is that the system init system will watch for the Kubelet POSIX lock and start the system Kubelet if the lock is missing. Once the system Kubelet starts it will launch the self-hosted Kubelet.
+
```

Comment:
- In the case where self-hosted is not running and the lock file doesn't exist (e.g. newly provisioned machine or a cold boot):
IIUC, when the system Kubelet is starting it is because the lock file didn't exist. After launching the system Kubelet, the lock is created. It isn't until there is contention detected on the lock detected by the system Kubelet that will use this as a signal to exit. When the system Kubelet exits, this creates a window of time for the self-hosted Kubelet to start and attempt to grab ahold of the lock itself.
- In the case where _any_ type of Kubelet was running, the lock file exists, but the Kubelet crashed:
A lock watchdog, outside of the Kubelet's watching mechanism, has to keep tabs on both the Kubelet process and the lock to associate that the lack of a running Kubelet means the lock should not exist. IMO, this functionality is best suited for the system init system to handle.

Thoughts?